File: OneCodeTeam/How-to-use-AlertTask-method-VBSharePointAlertTask-19804/[VB]-How to use AlertTask method (VBSharePointAlertTask)/VB/VBSharePointAlertTask/MyAlertTaskWF/MyAlertTaskWF.vb
'****************************** Module Header ******************************\
' Module Name:    MyAlertTaskWF.vb
' Project:        VBSharePointAlertTask
' Copyright (c) Microsoft Corporation
'
' This sample code will demo how to use AlertTask method to deal with a workflow task.
' 
' This source is subject to the Microsoft Public License.
' See http://www.microsoft.com/en-us/openness/licenses.aspx#MPL.
' All other rights reserved.
' 
' THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND, 
' EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE IMPLIED 
' WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.
'*****************************************************************************/
Imports System.ComponentModel
Imports System.ComponentModel.Design
Imports System.Drawing
Imports System.Workflow.ComponentModel.Compiler
Imports System.Workflow.ComponentModel.Serialization
Imports System.Workflow.ComponentModel
Imports System.Workflow.ComponentModel.Design
Imports System.Workflow.Runtime
Imports System.Workflow.Activities
Imports System.Workflow.Activities.Rules
Imports Microsoft.SharePoint.Workflow
Imports Microsoft.SharePoint.WorkflowActions

Public Class MyAlertTaskWF
    Inherits SequentialWorkflowActivity

    Public workflowProperties As New SPWorkflowActivationProperties

    Public Sub New()
        MyBase.New()
        InitializeComponent()
    End Sub

    ' workflow Id
    Public workflowId As New Guid

#Region "Properties"

    ' reassignee
    Private strReassignee As String = String.Empty
    ' flag for reassign
    Private isReassign As Boolean = False
    ' assignee
    Private assignee As String = Nothing
    ' init taskStatus
    Private strTaskStatus As String = "Pending Approval"

    'Flag for while.
    Private isFinished As Boolean = False
#End Region

    ''' <summary>
    ''' CreateTask Created
    ''' </summary>
    Private Sub createTask1_MethodInvoking(ByVal sender As Object, ByVal e As EventArgs)
        ' Generate new GUID used to initialize task correlation token 
        TaskId = Guid.NewGuid()

        If strReassignee IsNot Nothing AndAlso isReassign = True Then
            TaskProperties.AssignedTo = strReassignee
            TaskProperties.Title = String.Format("Review 2:【{0}】", Me.workflowProperties.Item.DisplayName)
        Else
            TaskProperties = New SPWorkflowTaskProperties()
            ' Assign initial properties prior to task creation 
            TaskProperties.Title = String.Format("Review:【{0}】", Me.workflowProperties.Item.DisplayName)
            TaskProperties.PercentComplete = 0
            TaskProperties.StartDate = DateTime.Today
            TaskProperties.DueDate = DateTime.Today.AddDays(3)
            TaskProperties.ExtendedProperties("TaskStatus") = strTaskStatus
            TaskProperties.AssignedTo = assignee
        End If

    End Sub

    'Condition of while
    Private Sub while1Invoke(ByVal sender As Object, ByVal e As ConditionalEventArgs)
        e.Result = Not isFinished
    End Sub

    ' Handler of OnTaskChanged
    Private Sub onTaskChanged_Invoked(ByVal sender As Object, ByVal e As ExternalDataEventArgs)
        Try
            isReassign = Boolean.Parse(TaskAfterProperties.ExtendedProperties("isReassign").ToString())
            If isReassign = True Then
                strReassignee = TaskAfterProperties.ExtendedProperties("ReAssignedTo").ToString()
                isFinished = False
            Else
                If TaskAfterProperties.ExtendedProperties("TaskStatus").ToString() = "Canceled" Then
                    isFinished = False
                Else
                    isFinished = (TaskAfterProperties.ExtendedProperties("TaskStatus").ToString() = "Completed")
                End If
            End If
        Catch ex As Exception

            logForTaskChanged.HistoryDescription = ex.ToString()
        End Try

    End Sub

    Private Sub logTaskComplete_MethodInvoking(ByVal sender As Object, ByVal e As EventArgs)
        HistoryDescription = "TaskOutcome: " & TaskAfterProperties.ExtendedProperties("TaskOutcome").ToString() & ";AssignedTo:" & Convert.ToString(TaskProperties.AssignedTo)

        HistoryOutcome = "Task Completed"
    End Sub

    Private Sub logForTaskChanged_MethodInvoking(ByVal sender As Object, ByVal e As EventArgs)
        HistoryDescription = "TaskOutcome: " & TaskAfterProperties.ExtendedProperties("TaskOutcome").ToString() & ";AssignedTo:" & Convert.ToString(TaskProperties.AssignedTo)

        HistoryOutcome = "Task Changed"
    End Sub

    Private Sub logActivated_MethodInvoking(ByVal sender As Object, ByVal e As EventArgs)
        HistoryDescription = "Workflow data: " & "Approver=seiya"
        HistoryOutcome = "Workflow activated"
    End Sub

    Private Sub logTaskCreated_MethodInvoking(ByVal sender As Object, ByVal e As EventArgs)
        HistoryDescription = "Task data: " & "AssignedTo=" & Convert.ToString(TaskProperties.AssignedTo) & "; " & "TaskStatus=" & TaskProperties.ExtendedProperties("TaskStatus").ToString() & "; " & "TaskTitle=" & Convert.ToString(TaskProperties.Title)
        HistoryOutcome = "Task created"
    End Sub

    Private Sub onWFActivated_Invoked(ByVal sender As Object, ByVal e As ExternalDataEventArgs)
        Try
            workflowId = workflowProperties.WorkflowId

            assignee = "seiya"
        Catch ex As Exception
            logActivated.HistoryDescription = ex.ToString()
        End Try
    End Sub
    Public Shared HistoryDescriptionProperty As DependencyProperty = DependencyProperty.Register("HistoryDescription", GetType(System.String), GetType(VBSharePointAlertTask.MyAlertTaskWF))

    <DesignerSerializationVisibilityAttribute(DesignerSerializationVisibility.Visible)> _
        <BrowsableAttribute(True)> _
            <CategoryAttribute("Misc")> _
    Public Property HistoryDescription As System.String
        Get
            Return CType(MyBase.GetValue(VBSharePointAlertTask.MyAlertTaskWF.HistoryDescriptionProperty), String)

        End Get
        Set(ByVal value As System.String)
            MyBase.SetValue(VBSharePointAlertTask.MyAlertTaskWF.HistoryDescriptionProperty, value)

        End Set
    End Property
    Public Shared HistoryOutcomeProperty As DependencyProperty = DependencyProperty.Register("HistoryOutcome", GetType(System.String), GetType(VBSharePointAlertTask.MyAlertTaskWF))

    <DesignerSerializationVisibilityAttribute(DesignerSerializationVisibility.Visible)> _
        <BrowsableAttribute(True)> _
            <CategoryAttribute("Misc")> _
    Public Property HistoryOutcome As System.String
        Get
            Return CType(MyBase.GetValue(VBSharePointAlertTask.MyAlertTaskWF.HistoryOutcomeProperty), String)

        End Get
        Set(ByVal value As System.String)
            MyBase.SetValue(VBSharePointAlertTask.MyAlertTaskWF.HistoryOutcomeProperty, value)

        End Set
    End Property
    Public Shared TaskIdProperty As DependencyProperty = DependencyProperty.Register("TaskId", GetType(System.Guid), GetType(VBSharePointAlertTask.MyAlertTaskWF))

    <DesignerSerializationVisibilityAttribute(DesignerSerializationVisibility.Visible)> _
        <BrowsableAttribute(True)> _
            <CategoryAttribute("Misc")> _
    Public Property TaskId As System.Guid
        Get
            Return CType(MyBase.GetValue(VBSharePointAlertTask.MyAlertTaskWF.TaskIdProperty), System.Guid)

        End Get
        Set(ByVal value As System.Guid)
            MyBase.SetValue(VBSharePointAlertTask.MyAlertTaskWF.TaskIdProperty, value)

        End Set
    End Property
    Public Shared TaskPropertiesProperty As DependencyProperty = DependencyProperty.Register("TaskProperties", GetType(Microsoft.SharePoint.Workflow.SPWorkflowTaskProperties), GetType(VBSharePointAlertTask.MyAlertTaskWF))

    <DesignerSerializationVisibilityAttribute(DesignerSerializationVisibility.Visible)> _
        <BrowsableAttribute(True)> _
            <CategoryAttribute("Misc")> _
    Public Property TaskProperties As SPWorkflowTaskProperties
        Get
            Return CType(MyBase.GetValue(VBSharePointAlertTask.MyAlertTaskWF.TaskPropertiesProperty), Microsoft.SharePoint.Workflow.SPWorkflowTaskProperties)

        End Get
        Set(ByVal value As SPWorkflowTaskProperties)
            MyBase.SetValue(VBSharePointAlertTask.MyAlertTaskWF.TaskPropertiesProperty, value)

        End Set
    End Property
    Public Shared TaskAfterPropertiesProperty As DependencyProperty = DependencyProperty.Register("TaskAfterProperties", GetType(Microsoft.SharePoint.Workflow.SPWorkflowTaskProperties), GetType(VBSharePointAlertTask.MyAlertTaskWF))

    <DesignerSerializationVisibilityAttribute(DesignerSerializationVisibility.Visible)> _
        <BrowsableAttribute(True)> _
            <CategoryAttribute("Misc")> _
    Public Property TaskAfterProperties As SPWorkflowTaskProperties
        Get
            Return CType(MyBase.GetValue(VBSharePointAlertTask.MyAlertTaskWF.TaskAfterPropertiesProperty), Microsoft.SharePoint.Workflow.SPWorkflowTaskProperties)

        End Get
        Set(ByVal value As SPWorkflowTaskProperties)
            MyBase.SetValue(VBSharePointAlertTask.MyAlertTaskWF.TaskAfterPropertiesProperty, value)

        End Set
    End Property
End Class

File: OneCodeTeam/How-to-use-AlertTask-method-CSSharePointAlertTask-19803/[C#]-How to use AlertTask method (CSSharePointAlertTask)/C#/CSSharePointAlertTask/MyAlertTaskWF/MyAlertTaskWF.cs
/****************************** Module Header ******************************\
* Module Name:    MyAlertTaskWF.cs
* Project:        CSSharePointAlertTask
* Copyright (c) Microsoft Corporation
*
* This sample code will demo how to use AlertTask method to deal with a workflow task.
* 
* This source is subject to the Microsoft Public License.
* See http://www.microsoft.com/en-us/openness/licenses.aspx#MPL.
* All other rights reserved.
* 
* THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND, 
* EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE IMPLIED 
* WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.
\*****************************************************************************/

using System;
using System.ComponentModel;
using System.Workflow.ComponentModel;
using System.Workflow.Activities;
using Microsoft.SharePoint.Workflow;

namespace CSSharePointAlertTask.MyAlertTaskWF
{
    public sealed partial class MyAlertTaskWF : SequentialWorkflowActivity
    {
        public MyAlertTaskWF()
        {
            InitializeComponent();
        }

        public Guid workflowId = default(System.Guid);
        public SPWorkflowActivationProperties workflowProperties = new SPWorkflowActivationProperties();

        #region Properties

        // reassignee
        string strReassignee = string.Empty;
        // flag for reassign
        private bool isReassign = false;
        // assignee
        string assignee = default(System.String);
        // init taskStatus
        string strTaskStatus = "Pending Approval";
        // Flag for while.
        private bool isFinished = false;
        #endregion

        /// <summary>
        /// CreateTask Created
        /// </summary>
        private void createTask1_MethodInvoking(object sender, EventArgs e)
        {
            // Generate new GUID used to initialize task correlation token 
            TaskId = Guid.NewGuid();

            if (strReassignee != null && isReassign == true)
            {
                TaskProperties.AssignedTo = strReassignee;
                TaskProperties.Title = string.Format("Review 2:【{0}】", this.workflowProperties.Item.DisplayName);
            }
            else
            {
                TaskProperties = new SPWorkflowTaskProperties();
                // Assign initial properties prior to task creation 
                TaskProperties.Title = string.Format("Review:【{0}】", this.workflowProperties.Item.DisplayName);
                TaskProperties.PercentComplete = 0;
                TaskProperties.StartDate = DateTime.Today;
                TaskProperties.DueDate = DateTime.Today.AddDays(3);
                TaskProperties.ExtendedProperties["TaskStatus"] = strTaskStatus;
                TaskProperties.AssignedTo = assignee;
            }

        }

        // Condition of while
        private void while1Invoke(object sender, ConditionalEventArgs e)
        {
            e.Result = !isFinished;
        }

        // Handler of OnTaskChanged
        private void onTaskChanged_Invoked(object sender, ExternalDataEventArgs e)
        {
            try
            {
                isReassign = bool.Parse(TaskAfterProperties.ExtendedProperties["isReassign"].ToString());
                if (isReassign == true)
                {
                    strReassignee = TaskAfterProperties.ExtendedProperties["ReAssignedTo"].ToString();
                    isFinished = false;
                }
                else
                {
                    if (TaskAfterProperties.ExtendedProperties["TaskStatus"].ToString() == "Canceled")
                    {
                        isFinished = false;
                    }
                    else
                    {
                        isFinished = (TaskAfterProperties.ExtendedProperties["TaskStatus"].ToString() == "Completed");
                    }
                }
            }
            catch (Exception ex)
            {

                logForTaskChanged.HistoryDescription = ex.ToString();
            }

        }

        public static DependencyProperty TaskIdProperty = DependencyProperty.Register("TaskId", typeof(System.Guid), typeof(CSSharePointAlertTask.MyAlertTaskWF.MyAlertTaskWF));

        [DesignerSerializationVisibilityAttribute(DesignerSerializationVisibility.Visible)]
        [BrowsableAttribute(true)]
        [CategoryAttribute("Misc")]
        public Guid TaskId
        {
            get
            {
                return ((System.Guid)(base.GetValue(CSSharePointAlertTask.MyAlertTaskWF.MyAlertTaskWF.TaskIdProperty)));
            }
            set
            {
                base.SetValue(CSSharePointAlertTask.MyAlertTaskWF.MyAlertTaskWF.TaskIdProperty, value);
            }
        }

        public static DependencyProperty TaskPropertiesProperty = DependencyProperty.Register("TaskProperties", typeof(Microsoft.SharePoint.Workflow.SPWorkflowTaskProperties), typeof(CSSharePointAlertTask.MyAlertTaskWF.MyAlertTaskWF));

        [DesignerSerializationVisibilityAttribute(DesignerSerializationVisibility.Visible)]
        [BrowsableAttribute(true)]
        [CategoryAttribute("Misc")]
        public SPWorkflowTaskProperties TaskProperties
        {
            get
            {
                return ((Microsoft.SharePoint.Workflow.SPWorkflowTaskProperties)(base.GetValue(CSSharePointAlertTask.MyAlertTaskWF.MyAlertTaskWF.TaskPropertiesProperty)));
            }
            set
            {
                base.SetValue(CSSharePointAlertTask.MyAlertTaskWF.MyAlertTaskWF.TaskPropertiesProperty, value);
            }
        }

        public static DependencyProperty TaskAfterPropertiesProperty = DependencyProperty.Register("TaskAfterProperties", typeof(Microsoft.SharePoint.Workflow.SPWorkflowTaskProperties), typeof(CSSharePointAlertTask.MyAlertTaskWF.MyAlertTaskWF));

        [DesignerSerializationVisibilityAttribute(DesignerSerializationVisibility.Visible)]
        [BrowsableAttribute(true)]
        [CategoryAttribute("Misc")]
        public SPWorkflowTaskProperties TaskAfterProperties
        {
            get
            {
                return ((Microsoft.SharePoint.Workflow.SPWorkflowTaskProperties)(base.GetValue(CSSharePointAlertTask.MyAlertTaskWF.MyAlertTaskWF.TaskAfterPropertiesProperty)));
            }
            set
            {
                base.SetValue(CSSharePointAlertTask.MyAlertTaskWF.MyAlertTaskWF.TaskAfterPropertiesProperty, value);
            }
        }

        public static DependencyProperty HistoryDescriptionProperty = DependencyProperty.Register("HistoryDescription", typeof(System.String), typeof(CSSharePointAlertTask.MyAlertTaskWF.MyAlertTaskWF));

        [DesignerSerializationVisibilityAttribute(DesignerSerializationVisibility.Visible)]
        [BrowsableAttribute(true)]
        [CategoryAttribute("Misc")]
        public String HistoryDescription
        {
            get
            {
                return ((string)(base.GetValue(CSSharePointAlertTask.MyAlertTaskWF.MyAlertTaskWF.HistoryDescriptionProperty)));
            }
            set
            {
                base.SetValue(CSSharePointAlertTask.MyAlertTaskWF.MyAlertTaskWF.HistoryDescriptionProperty, value);
            }
        }

        public static DependencyProperty HistoryOutcomeProperty = DependencyProperty.Register("HistoryOutcome", typeof(System.String), typeof(CSSharePointAlertTask.MyAlertTaskWF.MyAlertTaskWF));

        [DesignerSerializationVisibilityAttribute(DesignerSerializationVisibility.Visible)]
        [BrowsableAttribute(true)]
        [CategoryAttribute("Misc")]
        public String HistoryOutcome
        {
            get
            {
                return ((string)(base.GetValue(CSSharePointAlertTask.MyAlertTaskWF.MyAlertTaskWF.HistoryOutcomeProperty)));
            }
            set
            {
                base.SetValue(CSSharePointAlertTask.MyAlertTaskWF.MyAlertTaskWF.HistoryOutcomeProperty, value);
            }
        }

        private void logTaskComplete_MethodInvoking(object sender, EventArgs e)
        {
            HistoryDescription = "TaskOutcome: " +
                                          TaskAfterProperties.ExtendedProperties["TaskOutcome"].ToString() + ";AssignedTo:" + TaskProperties.AssignedTo;

            HistoryOutcome = "Task Completed";
        }

        private void logForTaskChanged_MethodInvoking(object sender, EventArgs e)
        {
            HistoryDescription = "TaskOutcome: " +
                                          TaskAfterProperties.ExtendedProperties["TaskOutcome"].ToString() + ";AssignedTo:" + TaskProperties.AssignedTo;

            HistoryOutcome = "Task Changed";
        }

        private void logActivated_MethodInvoking(object sender, EventArgs e)
        {
            HistoryDescription = "Workflow data: " +
                                 "Approver=seiya";
            HistoryOutcome = "Workflow activated";
        }

        private void logTaskCreated_MethodInvoking(object sender, EventArgs e)
        {
            HistoryDescription = "Task data: " +
                                  "AssignedTo=" + TaskProperties.AssignedTo + "; " +
                                  "TaskStatus=" + TaskProperties.ExtendedProperties["TaskStatus"].ToString() + "; " +
                                  "TaskTitle=" + TaskProperties.Title;
            HistoryOutcome = "Task created";
        }

        private void onWFActivated_Invoked(object sender, ExternalDataEventArgs e)
        {
            try
            {
                workflowId = workflowProperties.WorkflowId;

                assignee = "seiya";
            }
            catch (Exception ex)
            {
                logActivated.HistoryDescription = ex.ToString();
            }
        }
    }
}
